[Subtitles] Show font family name's to subtitle fonts list

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This change allow to show in the subtitles font setting list, the font family name instead of the font filename, as happens on every text editor.
To achieve this has been introduced a new building fonts cache which is triggered when Kodi is started. This also solves the side effect of the GUI freeze every time you try to open the subtitle Kodi settings when there are hundreds/thousands of fonts,
and ensure to fix #20861 memory leak.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Initially started to fix issue #20861 for a memory leak,
i preferred to make more radical changes in order to improve the current architecture.

The memory leak was because `ass_set_fonts` was called more times,
and libass was not cleaningup the allocated memory, libass bug!?

The choice to call `ass_set_fonts` each time the style change was made for two main reasons:
- be able to set the selected font from our fonts folders (kodi system or user), this was a kind of workaround to manage fonts from more folders, thing not supported from libass (allow only one additional folder)
- was not implemented a way to get the font family name, kodi only provided the font filename
with the side effect in some cases to have the font provider disabled in libass.

Then implementing a way to retrieve the font family name,
allow to set the right font name to the libass style, avoid call `ass_set_fonts` each time (then no mem leak),
and allowing the advantage to have a font list with font family names in font settings.

To note that the default subtitle font, is now called as:
`Default DejaVu Sans`
this is a dynamic entry constructed by code, in two string parts:
"Default" + font family name, of the default kodi font file set by source code

Why the default Kodi font is not more called `Arial`?
As someone has tried to explain in this issue #19912, our current default Kodi font used in GUI and subtitles, is a real chimera,
the Arial name come from first Kodi versions (10.0 ?) but has nothing to do with Arial font, appears to be a merge of two differents fonts: DejaVu Sans + Droid Sans Fallback, no longer up-to-date.
After the future merge of GUI font rework PR, we are planning to remove this chimera Arial font

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Load more than 1000 fonts in user fonts folder
Load some SRT subtitle and look at loaded font name in libass log
Test an ASS subtitle for testing various overriding settings

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):
![immagine](https://user-images.githubusercontent.com/3257156/150825950-a1706f18-5d18-4e02-bd9b-435aa391f97f.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

mkreisl commented:
Build error (Debian 10 cross compile fpr Raspberry Pi4)
```
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp: In member function 'void CDVDSubtitlesLibass::Configure(
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:138:51: error: invalid conversion from 'const char*' to '
  138 |           ass_add_font(m_library, fontFamily.c_str(), reinterpret_cast<const char*>(buffer.data()),
      |                                   ~~~~~~~~~~~~~~~~^~
      |                                                   |
      |                                                   const char*
In file included from /home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.h:17,
                 from /home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:9:
/usr/include/ass/ass.h:696:47: note:   initializing argument 2 of 'void ass_add_font(ASS_Library*, char*, char*, int)'
  696 | void ass_add_font(ASS_Library *library, char *name, char *data,
      |                                         ~~~~~~^~~~
/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:138:55: error: invalid conversion from 'const char*' to '
  138 |           ass_add_font(m_library, fontFamily.c_str(), reinterpret_cast<const char*>(buffer.data()),
      |                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                       |
      |                                                       const char*
In file included from /home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.h:17,
                 from /home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:9:
/usr/include/ass/ass.h:696:59: note:   initializing argument 3 of 'void ass_add_font(ASS_Library*, char*, char*, int)'
  696 | void ass_add_font(ASS_Library *library, char *name, char *data,
      |                                                     ~~~~~~^~~~
make[2]: *** [build/cores/VideoPlayer/subtitles/CMakeFiles/dvdsubtitles.dir/build.make:186: build/cores/VideoPlayer/subtitles/CMakeFiles/dvdsubtitles.dir/DVDSubtitlesLibass
make[2]: Leaving directory '/home/xbian/xbian-package-xbmc/build/rpi4-nexus/working/build'
[ 87%] Building CXX object build/input/touch/generic/CMakeFiles/input_touch_generic.dir/GenericTouchInputHandler.cpp.o
make[1]: *** [CMakeFiles/Makefile2:10193: build/cores/VideoPlayer/subtitles/CMakeFiles/dvdsubtitles.dir/all] Error 2
```

lrusak commented:
> Build error (Debian 10 cross compile fpr Raspberry Pi4)
> 

Looks like this might require a newer version of libass. Ref: https://github.com/libass/libass/releases/tag/0.15.1

mkreisl commented:
WTF, of course this is not included in Debian 10 :angry:

mkreisl commented:
With libass 0.15.2 everything is fine

scott967 commented:
When I run this in Windows x64 VS 2017 debugger, the playersetting for languages subtitles.fontname is greyed out/ not enabled.

CastagnaIT commented:
@scott967  likely that you havent updated kodi system folder with pr

CastagnaIT commented:
i have fixed the build problems, also to build with older libass versions (FreeBSD and Debian)
i think is now ready for review

mkreisl commented:
> i have fixed the build problems, also to build with older libass versions (FreeBSD and Debian)
> i think is now ready for review

Would not have been necessary. The way I appreciate @lrusak, the #if ... #else ... #endif constructions are a thorn in his side. :smile:

enen92 commented on system/settings/settings.xml:
```diff
@@ -517,26 +517,24 @@
           <level>1</level>
-          <default>arial.ttf</default>
+          <!-- KodiDefault, is a placeholder font family name for the default Kodi font -->
+          <default>KodiDefault</default>
```
I am unsure how seriously we want to take rebranding but probably it is better to avoid the Kodi naming. Maybe SystemDefault, AppDefault or SubtitleDefault would work?

CastagnaIT replied:
good point i had forgotten about rebranding...
i think i will do as on `subtitles.charset` with capitals DEFAULT

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -37,19 +42,15 @@ constexpr const char* userFontPath = "special://home/media/Fonts/";
 constexpr const char* systemFontPath = "special://xbmc/media/Fonts/";
 
-std::string GetDefaultFontPath(std::string& font)
+std::string GetSystemFontPath(std::string font)
```
Maybe adjust comment on line 42, to refer the font families or the actual filenames

enen92 replied:
Any reason for removing the ref? Could be const ref here

CastagnaIT replied:
i have removed ref due to the call with FONT::FONT_DEFAULT_FILENAME
but i think i can change std::string with const char*

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -37,19 +42,15 @@ constexpr const char* userFontPath = "special://home/media/Fonts/";
-    {
-      return CSpecialProtocol::TranslatePath(fontPath).c_str();
-    }
+    return CSpecialProtocol::TranslatePath(fontPath).c_str();
```
You're returning cstr here but the signature of the method returns STD::string. Is the conversion intended?

CastagnaIT replied:
forgot thanks

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -109,9 +110,55 @@ void CDVDSubtitlesLibass::Configure()
+      const std::string filepath = item->GetPath();
+      if (file.LoadFile(filepath, buffer) > 0)
+      {
+        std::string fontFamily = FONT::GetFontFamily(buffer);
```
Why not turn this into GetFontFamily(path, buffer)? Load file should probably be done in utils

CastagnaIT replied:
i am not sure i have understand
do you want to suggest to add in the GetFontFamily an OUT "buffer" argument?

enen92 replied:
Yes, in-out by passing it as reference. Create it here and pass it to the method. Less bloat while the main logic is still in the fontutils scope

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -109,9 +110,55 @@ void CDVDSubtitlesLibass::Configure()
+    CLog::Log(LOGERROR,
+              "{} - The Kodi font {} is missing. The default subtitle font cannot be set.",
+              __FUNCTION__, FONT::FONT_DEFAULT_FILENAME);
+    m_defaultFontFamilyName = "KodiFontMissing";
```
This could probably be an empty string? Maybe clear() or even do nothing. Libass is able to cope with this as far as I remember. Adding some random string might cause confusion

CastagnaIT replied:
> Libass is able to cope with this as far as I remember

you mean that in case of empty string set on style->FontName libass auto fallback? never tried i will do a try

enen92 replied:
Yeap I think it will

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -316,105 +364,106 @@ void CDVDSubtitlesLibass::ApplyStyle(const std::shared_ptr<struct KODI::SUBTITLE
-
-    if (m_subtitleType != NATIVE || subStyle->assOverrideStyles != OverrideStyles::POSITIONS)
+    free(style->FontName);
+    if (subStyle->fontName == FONT_PLACEHOLDER_FAMILYNAME)
```
Not a big fan of `FONT_PLACEHOLDER_FAMILYNAME` as it sounds like an hack, which this is not. Why not compare to m_defaultFontFamilyNamedirectly?

CastagnaIT replied:
how do you compare with m_defaultFontFamilyNamedirectly?
FONT_PLACEHOLDER_FAMILYNAME is needed to detect if the default kodi font is set in the setting

enen92 commented on xbmc/settings/SubtitlesSettings.cpp:
```diff
@@ -8,14 +8,48 @@
+  std::string filename = item->GetLabel();
+  // Font filename that start with "tmp.font." are excluded because
+  // are temporary fonts extracted by the demuxer from MKV files
+  if (!item->m_bIsFolder && CUtil::IsSupportedFontExtension(filename) &&
```
Maybe move to font utils too to wrap the comparison and create some single point of access?

enen92 replied:
Thinking a bit more about this, since now we have support for `ass_add_font` and we got rid of the support for two font folds. We can probably get rid of this and make the demuxer extract fonts to the temp folder just like before.
Fonts in temp/Fonts will then be added on initialization using `ass_add_font`. Not mandatory to change then, I can do a followup to get rid of this once the PR gets in

CastagnaIT replied:
I have just made this requested change locally
it does not matter even if my changes are deleted immediately afterwards,
however that would be wonderful

enen92 commented on xbmc/settings/SubtitlesSettings.cpp:
```diff
@@ -70,3 +104,54 @@ void CSubtitlesSettings::OnSettingChanged(const std::shared_ptr<const CSetting>&
+  const int itemsSize = items.Size();
+  for (int i = 0; i < itemsSize; ++i)
+  {
+    std::thread t1{&addFont, std::ref(vecFontMutex), items[i], std::ref(fontCollection)};
```
I am unsure about this. Someone else should take a look. Is JobManager not viable here?

lrusak replied:
I'm not really sure what our stance is on using `std::thread` directly outside of `CThread`

lrusak replied:
Also, as said on Slack, maybe it's just better to do this once at startup and if new fonts are added then restarting kodi is required.

CastagnaIT replied:
it would make sense to have a some sort of font cache,
i don't think i have enough kodi knowledge to integrate a cache management to be run at startup
also in how/where to store the list of font metadata

enen92 replied:
FontManager would be the place for that I guess

CastagnaIT replied:
removed threads and used JobManager on startup

enen92 commented on xbmc/settings/SubtitlesSettings.h:
```diff
@@ -11,13 +11,19 @@
 {
+// This is a placeholder to keep the fontname setting valid
+// even if the default Kodi font will be changed
+constexpr const char* FONT_PLACEHOLDER_FAMILYNAME = "KodiDefault";
```
See previous comment. Better to avoid Kodi naming

enen92 commented on xbmc/utils/FontUtils.cpp:
```diff
@@ -0,0 +1,58 @@
+/*
+ *  Copyright (C) 2005-2022 Team Kodi
```
2022-

CastagnaIT replied:
what means?

enen92 replied:
Those headers refer to the actual file not the program itself. So creation date is 2022 not 2005, and you can leave the other empty (==present)

garbear replied:
I think @enen92 means `Copyright (C) 2022 Team Kodi`. That's what I do for newly created files.

enen92 reviewed (commented):
I like the approach you did to get rid of the historical font handling in libass. Some comments, I have yet to runtime test this

thexai commented:
Maybe unrelated but Kodi crash here **only** in debug mode:

![Captura de pantalla 2022-01-23 213111](https://user-images.githubusercontent.com/58434170/150696944-f841341d-0a25-4edd-8be5-b905d3d984c1.png)

It would be great to we be able to fix this as it makes it very difficult to debug other issues.

lrusak commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -109,9 +110,55 @@ void CDVDSubtitlesLibass::Configure()
+      if (file.LoadFile(filepath, buffer) > 0)
+      {
+        std::string fontFamily = FONT::GetFontFamily(buffer);
+        if (!fontFamily.empty())
```
```suggestion
        if (fontFamily.empty())
          continue;
```

lrusak commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -109,9 +110,55 @@ void CDVDSubtitlesLibass::Configure()
+      std::vector<uint8_t> buffer;
+      XFILE::CFile file;
+      const std::string filepath = item->GetPath();
+      if (file.LoadFile(filepath, buffer) > 0)
```
```suggestion
      if (file.LoadFile(filepath, buffer) <= 0)
        continue;
```

lrusak commented on xbmc/utils/FontUtils.cpp:
```diff
@@ -0,0 +1,58 @@
+  FT_Init_FreeType(&m_library);
+  if (!m_library)
+  {
+    CLog::Log(LOGERROR, "{}: Unable to initialize freetype library", __func__);
```
```suggestion
    CLog::LogF(LOGERROR, "Unable to initialize freetype library");
```

lrusak commented on xbmc/utils/FontUtils.cpp:
```diff
@@ -0,0 +1,58 @@
+  }
+  else
+  {
+    CLog::Log(LOGERROR, "{}: Failed to process font memory buffer", __func__);
```
```suggestion
    CLog::LogF(LOGERROR, "Failed to process font memory buffer");
```

lrusak commented on xbmc/utils/FontUtils.cpp:
```diff
@@ -0,0 +1,58 @@
+
+  if (file.LoadFile(filepath, buffer) <= 0)
+  {
+    CLog::Log(LOGERROR, "{}: Failed to load file {}", __func__, filepath);
```
```suggestion
    CLog::LogF(LOGERROR, "Failed to load file {}", filepath);
```

lrusak commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -109,9 +110,55 @@ void CDVDSubtitlesLibass::Configure()
+  {
+    // This should never happen and in this case the name of the font no longer
+    // matters because in any case will be choosen another fallback font
+    CLog::Log(LOGERROR,
```
```suggestion
    CLog::LogF(LOGERROR,
```
and remove the `__FUNCTION__`

lrusak reviewed (commented):
Looks like a nice addition. I added a couple comments about style.

CastagnaIT commented:
> Maybe unrelated but Kodi crash here only in debug mode:

are you sure to have copied the needed files for libass debug in kodi directory? otherwise it will crash always

CastagnaIT committed:
[GUI][Settings] Do not disable list setting with single item

Problem 1:
If the setting list has a single item the setting was disabled, the user do not
understand why the setting is disabled and think to a bug.
Even if a setting has only one item in list, it should not be disabled,
and should be disabled only when has none item.

Problem 2:
Happens when a setting list has dynamic items and not fixed items, like subtitles fonts list.
E.g. initially you have a list with more items, then you select one
item, next time the items in list change with only one item available, the setting become
disabled with an empty value due to the non-existent item, you no longer have the chance to select the item available, then you are stuck with the previous setting value.

CastagnaIT commented:
Resolved some requests,
i have not touched threads and font loading way ATM i am not sure how to deal with requests mentioned

@scott967 found the problem of disabled setting, by default kodi was disabling settings having single entry in list setting
I have noticed that this causes problems with dynamic items settings like this one,
this did not occur before because there were two fixed items (2 fonts) always shown

lrusak commented:
jenkins build this please

scott967 commented:
OK now with a default install I get the setting available with Default DejaVu Sans as only option which I guess is expected (doubt anyone wants teletext2).  But I added 2 user font ttf files from same family but different weight (regular in one and semi-bold in the other).  I just see the single select entry for the font family.  Is this expected?  When playing a subtitle, the regular weight is being used.  Also when I stop a video I am getting an exception thrown in ~CDVDSubtitlesLibass() in ass_renderer_done(m_renderer);
`Unhandled exception at 0x00007FF823F6F199 (ntdll.dll) in kodi.exe: 0xC0000374: A heap has been corrupted (parameters: 0x00007FF823FD77F0). occurred`

CastagnaIT commented:
>  But I added 2 user font ttf files from same family but different weight (regular in one and semi-bold in the other). I just see the single select entry for the font family. Is this expected?

usually if two files have same font family name are shown as one file, this happens also with text editors the weight will be managed internally

> Also when I stop a video I am getting an exception thrown in ~CDVDSubtitlesLibass() in ass_renderer_done(m_renderer);

if you have build kodi as debug you have to manually put in kodi directory the libass debug files (should be contained in the kodi package) otherwise will crash, or as alternative you have to manually build libass lib as debug

scott967 commented:
Thanks.  Works as expected building release.  Is good to go for me.

Suggest adding a little more detail on wiki subtitles page 5.1 [(https://kodi.wiki/view/Subtitles)](https://kodi.wiki/view/Subtitles)

CastagnaIT commented:
I have integrated user fonts cache building at Kodi start-up, and deleted the two threads workaroud to half loading times in the hope of using in the future the parallels c++ features
@enen92 i think more or less this is what you meant

enen92 commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -508,19 +536,139 @@ void GUIFontManager::SettingOptionsFontsFiller(const SettingConstPtr& setting,
+
+void GUIFontManager::Initialize()
+{
+  m_isUpdating = true;
```
Why not a lock/mutex?

CastagnaIT replied:
mutex added

CastagnaIT committed:
[StringUtils] Add const to sortstringbyname operator

enen92 commented on xbmc/Application.cpp:
```diff
@@ -640,6 +640,29 @@ bool CApplication::Initialize()
   CServiceBroker::GetRenderSystem()->ShowSplash("");
 
+  // Initialize GUI font manager to build/update fonts cache
+  //!@todo Move GUIFontManager in to service broker and drop the global reference
```
```suggestion
  //! @todo Move GUIFontManager into service broker and drop the global reference
```

enen92 commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -1754,11 +1755,16 @@ CDemuxStream* CDVDDemuxFFmpeg::AddStream(int streamIdx)
+            // we will also use this folder to store fonts extracted by the
+            // demuxer. The extracted fonts will have a prefix in the filename
+            // for easy identification.
+            //!@todo: this font file management system on disk could be completely
```
```suggestion
            //! @todo: this font file management system on disk could be completely
```

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -32,24 +37,20 @@ constexpr int ASS_BORDER_STYLE_SQUARE_BOX = 4; // Square box + outline
 constexpr int ASS_FONT_ENCODING_AUTO = -1;
 
-// Directory where user defined fonts are located (and where mkv fonts are extracted to)
+// Directory where user defined fonts are located (and where temporary MKV fonts are extracted to)
```
```suggestion
// Directory where user defined fonts are located (and where MKV fonts are extracted and temporary stored)
```
i.e. the fonts themselves are not temporary :)

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -32,24 +37,20 @@ constexpr int ASS_BORDER_STYLE_SQUARE_BOX = 4; // Square box + outline
 constexpr const char* systemFontPath = "special://xbmc/media/Fonts/";
 
-std::string GetDefaultFontPath(std::string& font)
+const char* GetSystemFontPath(const char* font)
```
would be nice to keep all generic methods with `std::string` instead of `char *`. I'll do it later on the tmp cleanup if you don't want to get bothered with it

enen92 commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:
```diff
@@ -32,24 +37,20 @@ constexpr int ASS_BORDER_STYLE_SQUARE_BOX = 4; // Square box + outline
   }
-  CLog::Log(LOGERROR, "CDVDSubtitlesLibass: Could not find font {} in font sources", font);
+
+  CLog::Log(LOGERROR, "CDVDSubtitlesLibass: Could not find font {} in Kodi system folder", font);
```
LogF so that CDVDSubtitlesLibass can be removed from the log line?

enen92 commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -508,19 +528,140 @@ void GUIFontManager::SettingOptionsFontsFiller(const SettingConstPtr& setting,
+  previousCacheSize = m_userFontsCache.size();
+
+  // Add new files in cache and generate the metadata
+  //!@todo FIXME: this "for" loop should be replaced with the more performant
```
```suggestion
  //! @todo FIXME: this "for" loop should be replaced with the more performant
```

enen92 commented on xbmc/guilib/GUIFontManager.h:
```diff
@@ -40,6 +42,17 @@ struct OrigFontInfo
+struct FontMetadata
+{
+  FontMetadata(const std::string& filename, const std::string& familyName)
+    : m_filename(filename), m_familyName(familyName)
```
Out of curiosity, does default initialization with {} work on these cases? I mean `: m_filename{filename}, m_familyName{familyName}`? I suppose it doesn't

CastagnaIT replied:
yes should works, i have just read c++ guidelines

enen92 replied:
If it works we should use it, I always found that format of initialization for members in the base class extremely confusing as it looks like those members are functions or something similar

enen92 commented on xbmc/guilib/GUIFontManager.h:
```diff
@@ -50,6 +63,14 @@ class GUIFontManager : public IMsgTargetCallback
 
+  /*!
+   *  \brief Initialize the fonts manager.
+   *         Checks that fonts cache are up to date, otherwise update it.
```
```suggestion
   *  Checks that fonts cache are up to date, otherwise update it.
```

enen92 commented on xbmc/guilib/GUIFontManager.h:
```diff
@@ -81,6 +102,8 @@ class GUIFontManager : public IMsgTargetCallback
                                         std::string& current,
                                         void* data);
 
+  std::vector<std::string> GetUserFontsFamilyNames();
```
would be nice to have docs for this one someone not familiar with the codebase won't understand what this does from the signature

CastagnaIT replied:
i have deleted the doc for mistake thanks

enen92 commented on xbmc/threads/Lockables.h:
```diff
@@ -50,6 +51,12 @@ namespace XbmcThreads
+     * \brief Check if have a lock owned
+     * \return True if have a lock, otherwise false
+     */
+    inline bool isLocked() const { return count > 0; }
```
Shouldn't this be `IsLocked()` ?

enen92 reviewed (approved):
Left some really minor comments. I also gave this a spin and works as expected, awesome work!
If all goes well, I plan to merge this next weekend. Will work in the meantime into getting rid of the `tmp.font*` hack we have for mkvs on top of this.

enen92 commented:
No idea why Jenkins is complaining for clang-format changes in `Lockables.h` when you simply added 2 new lines. I'd ignore those requests if they are 100% contained within that file. Any automatic format will just blow the diff

CastagnaIT commented:
@enen92 it is because the spaces are wrong in whole file

enen92 commented:
> @enen92 it is because the spaces are wrong in whole file

Please ignore those requests then :)

CastagnaIT committed:
[Threads] Add isLocked to get lock status

enen92 commented:
@CastagnaIT the cleanup for the old `tmp.font` hack is in https://github.com/enen92/xbmc/commit/fa10498bf2d20e5a2102a10d5c3d3e295e285a83 you can review and cherry-pick if interested, or we just wait until this one gets in and I'll PR later

CastagnaIT commented:
your intention is a kind of cleanup to avoid save the fonts to the user fonts folder,
could be good to avoid mixing temporary files with user files,
but the current embedded font support for MKV still to be an inefficent workaroud of implementation, and this does not solve it

currently we are doing a lot of operations not needed that cause of only delay:
save each font file on the hdd, find the fonts, open fonts files, load fonts, delete fonts...
where the only operation to do should be route the existing fonts data already in memory to `ass_add_font`

enen92 commented:
Yeap you're right, this is still inefficient as the data is already in memory. A different solution would require coupling between components - and as a result it has to be well though out. Maybe a possible solution would be to move the data to the `FontManager` in the demuxer and load it (move it again) from the fontmanager to the libasshandler. Using move would guarantee the data is not left in memory after usage as long as there's always a consumer.

CastagnaIT commented:
not bad as idea use FontManager to store embedded fonts, i will try to do something

enen92 commented:
I've already "looked" into it trying to picture something that could be extendable, abstract and benefit Kodi as whole, unfortunately it involves a bit of effort and small steps:

- Kill `g_FontManager` global, move it to the service broker
- Ideally some other component should be the parent of fontmanager, something similar to what we have for GUI (we have GUIComponent that is the parent of some other GUI related objects). This helps making building blocks and get a better picture of the application as a whole. Since Fonts are resources and might be available even if kodi is running headless (e.g. you could have subtitles in the webinterface using the same fonts you'd have in the GUI player), I'd go with `CResources` or something similar. `CResources` would hold the FontManager and initialize it. It's initialized from CApp.
- Skins should be removed from CApp and moved to a `SkinManager` element that is a child of `CGUIComponent`
- There are a lot of components that can provide fonts. We could have some sort of abstraction `CFontProvider` as a sort of abstract base class. This abstract class registers on `FontManager` as a provider with a given ID on construction/initialization and de-registers itself on destruction/deinitialization. Virtual methods for FontLoading, FontData loading, etc.
- Skins are fontproviders, should use it as a base
- Font addons are fontproviders, should use it as a base
- In the future, script addons that provides skin elements can also be font providers solving one of the issues that have always affected addons (they can't use custom fonts)
- The mkv demuxer is also a font provider
- etc,

In this concrete example, the demuxer would derive from `FontProvider` and load (`std::move`) transitory/temporary font data to the font manager. The libass handler (`CDVDLibass`) would then use the broker to get the resource component, from there the fontmanager, and move the transitory font data again into memory.

Just some thoughts. And yeap, it involves quite a bit of work - feel free to disagree or propose something different. That's the reason I still think my commit above, while not solving this for the long term, is still an intermediary solution that would allow us to cleanup the `temp` hack in libass and fontmanager while allowing us to move forward with some other alternative in the future.

CastagnaIT commented:
yes you intend to refactor part of these components so not so small changes,
i don't have such a broad view now, therefore some points i have understood some others are obscure to me,
i will see to include here your cleanup/changes to use the fonts temporary folder

CastagnaIT committed:
[Subtitles] Show font family name's to subtitle fonts list

CastagnaIT commented:
i have add the last commit but i have discarded the proposed use of string_view,
this change would make sense if the related utils methods would be converted to use string_view
doing this would mess up the PR and I don't like add static_cast everywhere in code
a good conversion/clean-up to use the new string_view can be done on a separate PR

CastagnaIT committed:
[Subtitles] Make temporary fonts extract to and loaded from temp folder

Co-authored-by: enen92 <enen92@users.noreply.github.com>

enen92 commented:
No strong feelings about string_view you're right, it's quite ugly if all other methods expect strings (can be done later).
I'll go ahead and merge this. Thanks a lot for your work here, it's far better than what we used to have

monsoons commented:
Thank you for your effort. The memory leak is solved with the latest release.
Could you plz check a problem I found?
If I change the font for subtitles on my shield TV, only web fonts worked fine. The other fonts are just displayed as the default font. Same thing happens on Windows PC. But if I install the font on windows PC, then I can see it properly on Kodi. I've tested several fonts such as Google Noto Sans CJK(CHINA JAPANESE KOREAN), and other CJK fonts which worked properly in previous releases. I don't know how to install fonts on android devices, so I hope I can use the fonts without installation like before.

CastagnaIT commented:
@monsoons please open an new Issue and fullfill the template by specifing the steps to replicate your problem, and what type font and type of subtitle used etc..

monsoons commented:
> @monsoons please open an new Issue and fullfill the template by specifing the steps to replicate your problem, and what type font and type of subtitle used etc..

Thank you for your comment. I figured it out why this happens. It's because of font naming. If a font has "preferred family (Windows only)" name in advanced naming category and if "preferred family name" is not equal to "font family name" Kodi fails to show that font(just displays the default font). I edited the "preferred family name" using FontCreator app. Removing "preferred family name" or editing "preferred family name" equal to "font family name" solved the problem. Preferred family name in all platforms and in all language categories needs to be modified.

https://noto-website-2.storage.googleapis.com/pkgs/NotoSansCJKkr-hinted.zip

These Google free fonts are not displayed properly until I install it on windows system or edit the preferred family name.
If this info is not enough, I'll open an new issue. Thank you very much for solving the memory leak. Have a nice day!

CastagnaIT commented:
I think I understand i will try to find a solution

thexai commented:
Debug info OSD font style/size has changed from this merged I think. Maybe is need to change something here:

https://github.com/xbmc/xbmc/blob/cdb14802e55d6ad12085059419ce515844364d45/xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.cpp#L148-L150

Also seems unused variable here:
https://github.com/xbmc/xbmc/blob/cdb14802e55d6ad12085059419ce515844364d45/xbmc/cores/VideoPlayer/VideoRenderers/DebugRenderer.cpp#L132

CastagnaIT commented:
@thexai you're right i have completely forgotten the debug... i will fix thanks
